magda-typescript-common/src/registry/AuthorizedRegistryClient.ts

```diff
@@ -14,27 +14,41 @@ import buildJwt from "../session/buildJwt";
 import { IncomingMessage } from "http";
 import { Maybe } from "tsmonad";
 
-export interface AuthorizedRegistryOptions extends RegistryOptions {
+export interface AuthorizedRegistryUserIdAndJwtSecretOptions
+    extends RegistryOptions {
+    userId: string;
     jwtSecret: string;
-    userId?: string;
+    jwt?: never;
 }
 
+export interface AuthorizedRegistryJwtOptions extends RegistryOptions {
+    jwt: string;
+    userId?: never;
+    jwtSecret?: never;
+}
+
+export type AuthorizedRegistryOptions =
+    | AuthorizedRegistryUserIdAndJwtSecretOptions
+    | AuthorizedRegistryJwtOptions;
+
 export default class AuthorizedRegistryClient extends RegistryClient {
     protected options: AuthorizedRegistryOptions;
-    protected jwt: string | undefined;
+    protected jwt: string;
 
     constructor(options: AuthorizedRegistryOptions) {
         if (options.tenantId === undefined || options.tenantId === null) {
             throw Error("A tenant id must be defined.");
         }
 
-        if (options.jwtSecret === undefined || options.jwtSecret === null) {
-            throw Error("JWT secret must be defined.");
+        if (options.userId && options.jwtSecret === null) {
```

Comment:
when `this.jwt` presents, you didn't use `userId` or  `jwtSecret`, right?

Reply:
If `jwt` is defined, `userId` and `jwtSecret` should not be defined. 
If `userId` is defined, it guarantees `jwtSecret` is also defined. However, `jwtSecret` could be null.
Maybe I also needs to check if jwt is null.